designs/validations.md

```diff
@@ -0,0 +1,288 @@
+	* For information more tailored to feedback for the end user (like progress in slow validations), it will be left up to the framework callers to implement this. Based how logging is currently used to communicate this progress, the information is usually presented by "grouping" validations and not at the individual validation level. So the *usecases*, where validations are grouped and given semantic value, will be responsible to define these different groups and print logs in between them as necessary. This can be easily accomplished by composing multiple runners. Given a traditional logger might not be the best way to implement a presentation layer of this kind, this approach also allows to replace the logger with something else (in the future) without having to refactor any of the validation logic.
+
+#### Validation error meta information
+The existing `validations.ValidationResult` allows handlers to return a `Remediation` message. Given we want to just use the `error` interface to make it easy to register validations and that `Remediation` is empty for most `ValidationResult`'s , we chose to make this optional. Validations wanting to provide a possible remediation for a failure, can return an `error` that also implements the `Fixable` interface.
+
+```go
+type Fixable type {
+	Remediation() string
+}
+```
+
+#### Retrying
+Existing validation runners, like `framework.ClusterValidator`, incorporate retries into the runner itself. That might seem sensible, given a lot of validations might want to rely on retries to bypass transient errors. However, not only this would complicate the API to register validations but the retry logic will always be limited by the API itself (which is only addressable by making the API more complex).
+
+In this case we would leave this out of the framework and let each validation handler decide how to deal with retries.
+
+#### Framework implementation
+Given the framework logic is very much independent from the object being validated (`cluster.Spec`), we propose to use generics so the same framework can be used for different types. The complexity added is minimal and even in the case where the framework grows in a direction where generics don't make sense, moving back to a concrete type would be trivial. There are already places in the codebase where we envision make use of this:
+
+```go
+// Validation is the logic for a validation of a type O.
+type Validation[O any] func(ctx context.Context, obj O) error
+```
+
+In order to be able to prevent validations from modifying `obj` and detecting when if that happens, we add a constrain to `O` that allows us to `DeepCopy` the object so we can compare after running all validations:
```

Comment:
In what situation do we need to compare the objects after running all validations? It would only be during right, and as you mentioned above the validations should not be modifying the objects.

Reply:
> In what situation do we need to compare the objects after running all validations?

Every time the set of validations run. The goal is to make it very difficult if not impossible to abuse validations ensuring the have an appropriately narrow scope.